fix: count MFA addresses in CountActiveMultiFactorCredentials for code method

wewelll opened the pull request:
Attempt to fix https://github.com/ory/network/issues/409

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

This text will be included in the changelog. If applicable, include links to documentation or pieces of code.
If your change includes breaking changes please add a code block documenting the breaking change:

```
BREAKING CHANGES: This patch changes the behavior of configuration item `foo` to do bar. To keep the existing
behavior please do baz.
```
-->

## Related issue(s)

<!--
If this pull request

1. is a fix for a known bug, link the issue where the bug was reported in the format of `#1234`;
2. is a fix for a previously unknown bug, explain the bug and how to reproduce it in this pull request;
3. implements a new feature, link the issue containing the design document in the format of `#1234`;
4. improves the documentation, no issue reference is required.

Pull requests introducing new features, which do not have a design document linked are more likely to be rejected and take on average 2-8 weeks longer to
get merged.

You can discuss changes with maintainers either in the Github Discussions in this repository or
join the [Ory Chat](https://www.ory.sh/chat).
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR.

Please be aware that pull requests must have all boxes ticked in order to be merged.

If you're unsure about any of them, don't hesitate to ask. We're here to help!
-->

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ory/kratos?pullRequest=4318) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ory/kratos?pullRequest=4318) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ory/kratos?pullRequest=4318) it.</sub>

aeneasr reviewed (commented):
Looks really good! I guess the only question is if there are regression cases where this would break an existing set up. So for example:

1. Dev has set up code MFA
2. An identity exists where config does not have the address
3. MFA does no longer work for them

The best way to figure this out is [here](https://github.com/ory/kratos/blob/6eeeaa8629264d60a77125de767e04c9c70b0155/session/manager_http_test.go#L414) and [here](https://github.com/ory/kratos/blob/6eeeaa8629264d60a77125de767e04c9c70b0155/session/manager_http_test.go#L495). A regression case for code is [here](https://github.com/ory/kratos/blob/6eeeaa8629264d60a77125de767e04c9c70b0155/session/manager_http_test.go#L513-L516). More regression cases are [here](https://github.com/ory/kratos/blob/6eeeaa8629264d60a77125de767e04c9c70b0155/identity/credentials_migrate_test.go#L55).

Also it makes sense to add an end to end test in test/e2e to test the case you have: An identity schema with a field for code MFA that is not required, and an identity that is trying to sign in without  filling in the data (should not show 2fa), and one that is singing in with filling out the data (should show 2fa). With such a test we can ensure that this won't break in the future when someone changes something.

Thank you!

wewelll commented on selfservice/strategy/code/strategy.go:
```diff
@@ -138,11 +139,40 @@ func (s *Strategy) CountActiveFirstFactorCredentials(ctx context.Context, cc map
+	}
+
+	// If config is empty, return 0
+	if len(creds.Config) == 0 {
```
done

wewelll commented:
Thank you for the advice @aeneasr 
I have added a commit to be retro-compatible with existing configs. Can you confirm that it looks good to you ? 
I'm trying to add an e2e test now

wewelll commented:
It looks like a test is broken in `selfservice/strategy/password` but I did not change anything there. Where could it come from ?

aeneasr commented:
Yes, that's on master - I'll rebase this PR then it should pass

wewelll commented:
Thank you, all tests are passing now. 

Here is a summary of the changes on the tests: 
- all existing tests are still passing
- added some test cases on `CountActiveMultiFactorCredentials` for the old and new version of the code credential
- added new tests cases for `TestDoesSessionSatisfy`
- added a new e2e test suite for optional mfa with code

Let me know if there is something more to do !
